### ISSUE TITLE
support `hamlit` as alternative HAML implementation?

### ISSUE THREAD

**Issue by fnordfish** [padrino/padrino-framework#2115]
I was trying to replace "haml" with [hamlit](https://github.com/k0kubun/hamlit) in one of my padrino projects.

Hamlit is supposed to be a [limited](https://github.com/k0kubun/hamlit/blob/master/REFERENCE.md#limitations) drop in replacement. But somehow it does not work due to padrino's own management of template renders(?).

My fail-fast try looked like this:

```ruby
module Padrino::Helpers::OutputHelpers
  class HamlHandler
    def engine_matches?(block)
      eval('!!defined?(_hamlit_compiler1)', block.binding)
    end

    def capture_from_template(*args, &block)
      super.html_safe
    end
end
```

... which almost worked aside from tag helpers:

```html
&lt;div class=&#39;form-group&#39;&gt;
&lt;label class=&#39;col-lg-2 control-label&#39; for=&#39;email&#39;&gt;Email&lt;/label&gt;
&lt;div class=&#39;col-lg-10&#39;&gt;<input type="email" name="email" id="email" autofocus="autofocus" class="form-control" />&lt;/div&gt;
&lt;/div&gt;
```

Any ideas? (I also accept "Do not use hamlit - it won't work with the admin" is an answer :))

**Comment by nesquena** [comment]
You can see some of the work they are doing to support rails here https://github.com/k0kubun/hamlit/blob/master/lib/hamlit/rails_template.rb. The handler you added did seem to work, it's just that the html is escaped. 

Haven't ever used hamlit so I can't give much direction here other than to say with a but more tinkering I'm sure you could get it to work and if you did then consider making a plugin http://padrinorb.com/guides/generators/plugins/ so it's easier for the next developer :)

**Comment by fnordfish** [comment]
After reading that `hamlit` and `slim` both use `temple` as a template-engine framework, I've basically copied all slim stuff and renamed to hamlit (and removed the non-existing `disable_capture` option):

```ruby
module Padrino
  module Helpers
    module OutputHelpers
      ##
      # Handler for Hamlit templates.
      #
      class HamlitHandler < AbstractHandler
        ##
        # Returns true if the block is for Hamlit.
        #
        def engine_matches?(block)
          block.binding.eval('defined? __in_hamlit_template')
        end
      end
      OutputHelpers.register(:haml, HamlitHandler)
    end
  end
end

module Padrino
  module Rendering
    class HamlitOutputBuffer < Temple::Generators::StringBuffer
      define_options :buffer_class => 'SafeBuffer'

      def call(exp)
        [preamble, compile(exp), postamble].flatten.compact.join('; '.freeze)
      end

      def create_buffer
        "#{buffer} = #{options[:buffer_class]}.new"
      end

      def concat(str)
        "#{buffer}.safe_concat((#{str}))"
      end
    end

    class HamlitTemplate < Hamlit::Template
      include SafeTemplate

      def precompiled_preamble(locals)
        "__in_hamlit_template = true\n" << super
      end
    end
  end
end

Tilt.prefer(Padrino::Rendering::HamlitTemplate, :haml)

Padrino::Rendering.engine_configurations[:haml] = {
  :generator => Padrino::Rendering::HamlitOutputBuffer,
  :buffer => "@_out_buf",
  :use_html_safe => true,
}
```

**Comment by fnordfish** [comment]
Plugin-gist is here: https://gist.github.com/fnordfish/25098891d6337a083a4c0e23a7adfd5e

**Comment by nesquena** [comment]
Glad you got it working! As a bonus would be great if you send a PR here https://github.com/padrino/padrino-recipes but otherwise glad it's working.

**Comment by ujifgc** [comment]
Implemented in #2127
